### PR TITLE
fix(migrations): split redis migrations into up and teardown

### PR DESCRIPTION
**Pull request by nowNick**
### Summary

"UP" phase should only contain non-destructive operations. "TEARDOWN" (or "FINISH") phase should be used to change/delete data.

### Checklist

- [x] The Pull Request has tests
- [x] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/blob/main/README.md)
- [x] N/A ~~There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE~~

### Issue reference

KAG-4419
https://github.com/Kong/kong/issues/12978

### CODE REVIEW

File: kong/plugins/acme/migrations/003_350_to_360.lua [hunk 1]
```diff
@@ -1,41 +1,46 @@
...
-                      'server_name', config #> '{storage_config, redis, ssl_server_name}',
-                      'timeout', config #> '{storage_config, redis, timeout}',
-                      'database', config #> '{storage_config, redis, database}'
-                  ) || jsonb_build_object(
-                      'extra_options',
-                      jsonb_build_object(
-                          'scan_count', config #> '{storage_config, redis, scan_count}',
-                          'namespace', config #> '{storage_config, redis, namespace}'
-                      )
+      teardown = function(connector, _)
```
Comment (nowNick) [inline kong/plugins/acme/migrations/003_350_to_360.lua:3]: I moved entire `up` section to `teardown`. The reason for that is for `acme` plugin we're only renaming `auth` to `password` inside `storage_config.redis`.
Apart from that we're adding `extra_options` to `storage_config.redis`. 

Now that I think of it, we could split it into addition of `extra_options` and password in the `up` section and the `teardown` section would contain removal of `auth` field. 

However I wonder what that brings us 🤔 I guess I'm missing some context on the `up` and `teardown` phases (i.e. how they are used).

**Review by hanshuebner — APPROVED** [review]

**Review by samugi — APPROVED** [review]

**Comment by bungle** [comment]
@nowNick @locao @hanshuebner @samugi, I do want to double check if this is re-entrant or not. I feel it is not.

**Comment by StarlightIbuki** [comment]
+1 with @bungle . The `up` part seems ok to be executed multiple times but not after executing `tearup`. It will lose all the Redis setup. The commands need careful checks before it does the change. Please fix this and then backport it @nowNick.

**Comment by StarlightIbuki** [comment]
BTW is it ok for new nodes to see the old fields in the config object? It will happen when the teardown is not done.

**Comment by outsinre** [comment]
Just a quick through the code, this PR follows the official guideline to split update and add operations into up and teardown respectively.

However, it does not resolve the timeout issues reported here https://github.com/Kong/kong/issues/12978. It is quite common, nowadays, customers may have hundreds of rate-limiting instances.

Maybe, we should adopt @bungle 's suggestion on `shorthand_fields` like this https://github.com/Kong/kong/blob/49aa2332b219826410c46cf207c9bb31ce25785d/kong/plugins/session/schema.lua#L216.

**Comment by hanshuebner** [comment]
> However, it does not resolve the timeout issues reported here #12978. It is quite common, nowadays, customers may have hundreds of rate-limiting instances.

#12761 addresses the timeout issue in the sense that it makes the environment variable `KONG_PG_TIMEOUT` work in addition to the `--db-timeout` option.  It may be worthwhile to change the default database timeout for command line operations to a higher value.

> Maybe, we should adopt @bungle 's suggestion on `shorthand_fields` like this
> https://github.com/Kong/kong/blob/49aa2332b219826410c46cf207c9bb31ce25785d/kong/plugins/session/schema.lua#L216

Can you elaborate how `shorthand_fields` can help with the timeout issue?

**Comment by hanshuebner** [comment]
> Can you elaborate how shorthand_fields can help with the timeout issue?

@bungle explained:  We should generally avoid data migrations if possible as they are scary and failure prone.  In this case, the new version can work with the old configration values and `shorthand_fields`, so a data migration might not even be needed.

**Comment by bungle** [comment]
> BTW is it ok for new nodes to see the old fields in the config object? It will happen when the teardown is not done.

It is perfectly fine. New nodes will remove unknown fields on read.

**Comment by bungle** [comment]
> We should generally avoid data migrations if possible as they are scary and failure prone. In this case, the new version can work with the old configuration values and `shorthand_fields`, so a data migration might not even be needed.

Yes. This is a bit debatable. As some people feel that leaving old data in db is somekind of cruft that we need to get rid of, or that will cause some other problems. The old values will stay there only as long as entity is not `saved`, When entity is`saved` the old fields (and now unknown) are removed from (db). Meanwhile we can migrate them over to new fields (without even checking if new fields are there already, as I explained, they will be gone on save).

I also do agree that we need to get rid of this "cruft", but perhaps it is best done with major version. When we get rid of cruft like here in finish:
```
            UPDATE plugins
            SET config =
              config
                #- '{storage_config,redis,auth}'
                #- '{storage_config,redis,ssl_server_name}'
                #- '{storage_config,redis,scan_count}'
                #- '{storage_config,redis,namespace}'
            WHERE name = 'acme';
```

It is a destructive operation. Now only way to get back to older Kong version is to restore db. If we didn't do this at all, it would be just matter of switching version back. And it would work even if new fields already had data copied over, because again, old nodes remove unknown fields on read.

**Comment by bungle** [comment]
Migrations can go wrong in many unexpected ways. Especially data migrations. We have seen timeouts, processed get killed in a mid, network issues, postgres version issues (migration works on my machine, but not on customer version of pg), we have seen that we needed to rewrite migrations because of bugs and then we needed to somehow support also people who ran the buggy migration if possible etc.

**Pull request merged**